File: 0872-leaf-similar-trees/0872-leaf-similar-trees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    public void dfs(TreeNode root, List<Integer> leaf) {
        if (root == null)
            return ;
        if (root.left == null && root.right == null) {
            leaf.add(root.val);
            return ;
        }
        dfs(root.left, leaf);
        dfs(root.right, leaf);
    }
    
    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        List<Integer> seq1 = new ArrayList<>();
        List<Integer> seq2 = new ArrayList<>();
        
        dfs(root1, seq1);
        dfs(root2, seq2);
        return seq1.equals(seq2);
    }
}

File: 0872-leaf-similar-trees/0872-leaf-similar-trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        def dfs(root: Optional[TreeNode], leaves: list):
            if not root:
                return 
            if not root.left and not root.right:
                leaves.append(root.val)
                return
            dfs(root.left, leaves)
            dfs(root.right, leaves)
        
        tree1 = []
        tree2 = []
        dfs(root1, tree1)
        dfs(root2, tree2)
        return tree1 == tree2